src/coreclr/src/jit/lower.cpp

```diff
@@ -6365,3 +6314,189 @@ void Lowering::ContainCheckBitCast(GenTree* node)
+    TryCreateAddrMode(ind->Addr(), true);
+    if (!comp->codeGen->gcInfo.gcIsWriteBarrierStoreIndNode(ind))
+    {
+        LowerStoreIndir(ind);
+    }
+}
+
+//------------------------------------------------------------------------
+// LowerIndir: a common logic to lower IND load or NullCheck.
+//
+// Arguments:
+//    ind - the ind node we are lowering.
+//
+void Lowering::LowerIndir(GenTreeIndir* ind)
+{
+    assert(ind->OperIs(GT_IND, GT_NULLCHECK));
+    // Process struct typed indirs separately unless they are unused;
+    // they only appear as the source of a block copy operation or a return node.
+    if (!ind->TypeIs(TYP_STRUCT) || ind->IsUnusedValue())
+    {
+        // TODO-Cleanup: We're passing isContainable = true but ContainCheckIndir rejects
+        // address containment in some cases so we end up creating trivial (reg + offfset)
+        // or (reg + reg) LEAs that are not necessary.
+        TryCreateAddrMode(ind->Addr(), true);
+        ContainCheckIndir(ind);
+
+        if (ind->OperIs(GT_NULLCHECK) || ind->IsUnusedValue())
+        {
+            // A nullcheck is essentially the same as an indirection with no use.
+            // The difference lies in whether a target register must be allocated.
+            // On XARCH we can generate a compare with no target register as long as the addresss
+            // is not contained.
+            // On ARM64 we can generate a load to REG_ZR in all cases.
+            // However, on ARM we must always generate a load to a register.
+            // In the case where we require a target register, it is better to use GT_IND, since
+            // GT_NULLCHECK is a non-value node and would therefore require an internal register
+            // to use as the target. That is non-optimal because it will be modeled as conflicting
+            // with the source register(s).
+            // So, to summarize:
+            // - On ARM64, always use GT_NULLCHECK for a dead indirection.
+            // - On ARM, always use GT_IND.
+            // - On XARCH, use GT_IND if we have a contained address, and GT_NULLCHECK otherwise.
+            // In all cases, change the type to TYP_INT.
+            //
+            ind->gtType = TYP_INT;
+#ifdef TARGET_ARM64
+            bool useNullCheck = true;
+#elif TARGET_ARM
+            bool useNullCheck = false;
+#else  // TARGET_XARCH
+            bool useNullCheck = !ind->Addr()->isContained();
+#endif // !TARGET_XARCH
+
+            if (useNullCheck && ind->OperIs(GT_IND))
+            {
+                ind->ChangeOper(GT_NULLCHECK);
+                ind->ClearUnusedValue();
+            }
+            else if (!useNullCheck && ind->OperIs(GT_NULLCHECK))
+            {
+                ind->ChangeOper(GT_IND);
+                ind->SetUnusedValue();
+            }
+        }
+    }
+    else
+    {
+        // if `ADDR` node under `IND(struct(ADDR))` under `STORE_OBJ`
+        // is a complex one it could benefit from a not contained `LEA`.
```

Comment:
```suggestion
        // If the `ADDR` node under `STORE_OBJ(dstAddr, IND(struct(ADDR))` 
        // is a complex one it could benefit from an `LEA` that is not contained.
```